packages/@pollyjs/adapter-node-http/src/utils/transport-wrapper.js

```diff
@@ -0,0 +1,212 @@
+      `Cannot restore unpatched transport ${this.name}`,
+      this.isPatched()
+    );
+
+    this.transport.request = nativeRequestMapping.get(this.transport);
+    nativeRequestMapping.delete(this.transport);
+  }
+
+  getHttpResponseData(res) {
+    return new Promise(resolve => {
+      let stream;
+
+      switch (res.headers['content-encoding']) {
+        case 'gzip':
+        case 'compress':
+        case 'deflate':
+          stream = res.pipe(zlib.createUnzip());
+          delete res.headers['content-encoding'];
+          break;
+        default:
+          stream = res;
+      }
+
+      const resBuffer = [];
+
+      stream.on('data', chunk => resBuffer.push(chunk));
+
+      stream.on('end', () => {
+        const data = Buffer.concat(resBuffer).toString('utf8');
+
+        resolve(data);
+      });
+    });
+  }
+
+  async passthrough(req) {
+    const res = await new Promise(resolve => {
+      req.once('response', response => resolve(response));
+      req[NATIVE_END].call(req);
+    });
+
+    const data = await this.getHttpResponseData(res);
+
+    const { headers, statusCode } = res;
+
+    return {
+      headers,
+      statusCode,
+      data
+    };
+  }
+
+  async respond(req, response) {
+    Object.keys(req[LISTENERS]).forEach(eventName => {
+      const listeners = req[LISTENERS][eventName];
+
+      listeners.forEach(listener => req.on(eventName, listener));
+    });
+
+    const fakeSocket = { readable: false };
+    const msg = new http.IncomingMessage(fakeSocket);
+
+    msg.statusCode = response.statusCode;
+    req.emit('response', msg);
+
+    for (const h in response.headers) {
+      msg.headers[h] = response.headers[h];
+    }
+
+    let body = response.body;
+
+    if (typeof body === 'undefined') {
+      body = [];
+    }
+
+    msg.emit('data', Buffer.from(body));
+    msg.emit('end');
+
+    req.emit('prefinish');
+    req.emit('finish');
+  }
+
+  createRequestWrapper() {
+    const wrapper = this;
+    const { adapter, transport } = wrapper;
+    const nativeRequest = nativeRequestMapping.get(transport);
+
+    return (...args) => {
+      /**
+       * args could be (url, options, callback) or (options, callback) depending on nodejs version
+       * We don't need to wrap callback we can just emit 'response' event on ClientRequest
+       * and it's more reliable to collect request information from ClientRequest instance
+       * We don't need to store args
+       */
+
+      const req = nativeRequest.call(transport, ...args);
+
+      const chunks = [];
+
+      const originalWrite = req.write;
+
+      req.write = (chunk, encoding, callback) => {
+        // TODO : handle encoding
+        chunks.push(chunk);
+
+        originalWrite.call(req, chunk, encoding, callback);
+      };
+
+      // This can be called multiple times with the same request instance
+      // so make sure we only save the original `end` method.
+      req[NATIVE_END] = req[NATIVE_END] || req.end;
```

Comment:
In Node 6, the transport.request method was getting called multiple times with the same request instance and was going into an infinite loop since `NATIVE_END` was the patched `end` method on line 148.

Reply:
Good catch, i didn't think we would have a problem like that in here :)